nomenclature/codelist.py

```diff
@@ -588,6 +600,48 @@ def vars_kwargs(self, variables: List[str]) -> List[VariableCode]:
             if self[var].agg_kwargs and not self[var].skip_region_aggregation
         ]
 
+    def validate_units(self, unit_mapping) -> bool:
+        if invalid_units := [
+            (variable, unit, self.mapping[variable].unit)
+            for variable, unit in unit_mapping.items()
+            if variable in self.variables
+            and not set(unit) == set(self.mapping[variable].unit)
+        ]:
+            lst = [
+                f"'{v}' - expected: {'one of ' if isinstance(e, list) else ''}"
+                f"'{e}', found: '{u}'"
+                for v, u, e in invalid_units
+            ]
+            msg = "The following variable(s) are reported with the wrong unit:"
+            logging.error("\n - ".join([msg] + lst))
+            return False
+        return True
+
+    def validate_dataframe(self, df: IamDataFrame, dimension: str) -> bool:
+        # validate variables
+        all_variables_valid = super().validate_dataframe(df, dimension)
+        all_units_valid = self.validate_units(df.unit_mapping)
+        return all_variables_valid and all_units_valid
+
+    def add_missing_variables(
```

Comment:
This method doesn't actually add the variables, just prints them out in an easy-to-add format...
```suggestion
        def list_missing_variables(
```